controllers/etcdcluster_controller.go

```diff
@@ -0,0 +1,96 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	etcdv1alpha1 "github.com/improbable-eng/etcd-cluster-operator/api/v1alpha1"
+)
+
+// EtcdClusterReconciler reconciles a EtcdCluster object
+type EtcdClusterReconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+func headlessServiceForCluster(cluster *etcdv1alpha1.EtcdCluster) *v1.Service {
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(cluster, etcdv1alpha1.GroupVersion.WithKind("EtcdCluster")),
+			},
+			Labels: map[string]string{
+				appLabel:     appName,
+				clusterLabel: cluster.Name,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			ClusterIP:                v1.ClusterIPNone,
+			PublishNotReadyAddresses: true,
+			Selector: map[string]string{
+				appLabel:     appName,
+				clusterLabel: cluster.Name,
```

Comment:
The selector should be provided by the user through `etcdcluster.spec.selector` instead of being computed automatically - it makes a number of other areas easier in future. Can be left as-is for now, but I think we'll be changing this up 😄

Reply:
I don't think I understand the use-case for specifying it.

We're always going to want to add `app.kubernetes.io/name: etcd` and `etcd.improbable.io/cluster-name: $CLUSTERNAME`. So even if we had `etcdcluster.spec.selector` that would just be additive?

Plus if we did that we'd need to have an equivalent on the `EtcdPeer` so we could ensure it applies the same labels downstream too (so that the selectors actually function).